File: frontend/src/utils/validation.ts

export const isValidWebUrl = (urlString: string): boolean => {
    try {

        if (urlString === undefined || urlString === null)
            return false;

        if (urlString.length === 0)
            return false;

        const url = new URL(urlString);

        const hasWebProtocol = url.protocol === "http:" || url.protocol === "https:";
        const isNotLocal = url.hostname !== "localhost";
        const isNotLocalLookup = url.hostname !== "127.0.0.1";

        return hasWebProtocol && isNotLocal && isNotLocalLookup;

    } catch {
        return false;
    }
}

export const emailRules = [
    (value: string) => !!value || 'Email is required',
    (value: string) => /.+@.+\..+/.test(value) || 'Enter a valid email address',
]

export const usernameRules = [
    (value: string) => !!value || 'Username is required',
    (value: string) => value.length >= 1 || 'Username is required',
    (value: string) => value.length <= 255 || 'Username must be 255 characters or fewer',
]

export const passwordRules = [
    (value: string) => !!value || 'Password is required',
    (value: string) => value.length >= 8 || 'Password must be at least 8 characters',
    (value: string) => value.length <= 255 || 'Password must be 255 characters or fewer',
    (value: string) => /[A-Z]/.test(value) || 'Password must contain at least one uppercase letter',
    (value: string) => /[a-z]/.test(value) || 'Password must contain at least one lowercase letter',
    (value: string) => /\d/.test(value) || 'Password must contain at least one digit',
    (value: string) => /[^\w\s]/.test(value) || 'Password must contain at least one special character',
]

export const confirmPasswordRules = (password: string | null) => [
    (value: string) => !!value || 'Please confirm your password',
    (value: string) => value === password || 'Passwords do not match',
]
export const roleNameRules = [
    (value: string) => !!value || 'Role name is required',
    (value: string) => value.length >= 1 || 'Role name is required',
    (value: string) => value.length <= 255 || 'Role name must be 255 characters or fewer',
]
export const roleDescriptionRules = [
    (value: string) => !value || value.length <= 255 || 'Role description must be 255 characters or fewer',
]

export const permissionDescriptionRules = [
    (value: string) => (!!value && value.length <= 255) || 'Permission description must be 255 characters or fewer',
]

export const permissionResourceRules = [
    (value: string) => !!value || 'Permission resource is required',
    (value: string) => value.length <= 255 || 'Permission resource must be 255 characters or fewer',
    (value: string) => (['source', 'article', 'alert', 'job', 'keyword', 'dashboard'].includes(value)) || 'Permission resource must be one of: source, article, alert, job, keyword, dashboard',
]

export const permissionActionRules = [
    (value: string) => (!!value && value.length <= 255) || 'Permission action must be 255 characters or fewer',
    (value: string) => (['create', 'read', 'update', 'delete', 'run'].includes(value)) || 'Permission action can only contain letters, numbers, underscores, and hyphens',
]

export const permissionScopeRules = [
    (value: string) => (!!value && value.length <= 255) || 'Permission scope must be 255 characters or fewer',
    (value: string) => (['*', 'own', 'any'].includes(value)) || 'Permission scope can only contain letters, numbers, underscores, and hyphens',
]

File: backend/app/utils/validation.py
import re
from app.core.rbac import Resources, Actions, ScopeMode


def validate_password(password: str) -> None:
    """Validate password complexity requirements."""
    if not re.search(r"[A-Z]", password):
        raise ValueError("Password must contain at least one uppercase letter")
    if not re.search(r"[a-z]", password):
        raise ValueError("Password must contain at least one lowercase letter")
    if not re.search(r"\d", password):
        raise ValueError("Password must contain at least one digit")
    if not re.search(r"[^\w\s]", password):
        raise ValueError("Password must contain at least one special character")


def validate_resources(resource: str) -> None:
    """Validate resource name against allowed resources."""
    
    if resource is None:
        raise ValueError("Resource name cannot be None.")
    if not isinstance(resource, str):
        raise ValueError("Resource name must be a string.")
    if not Resources.has_value(resource):
        raise ValueError(f"Invalid resource name: {resource}")


def validate_actions(action: str) -> None:
    """Validate action name against allowed actions."""
    
    if action is None:
        raise ValueError("Action name cannot be None.")
    if not isinstance(action, str):
        raise ValueError("Action name must be a string.")
    if not Actions.has_value(action):
        raise ValueError(f"Invalid action name: {action}")
    
def validate_scope(scope: str) -> None:
    """Validate scope name against allowed scopes."""
    
    if scope is None:
        raise ValueError("Scope name cannot be None.")
    if not isinstance(scope, str):
        raise ValueError("Scope name must be a string.")
    if not ScopeMode.has_value(scope):
        raise ValueError(f"Invalid scope name: {scope}")
    